lxd/device/disk.go

```diff
@@ -552,6 +559,61 @@ func (d *disk) createDevice() (string, error) {
 	isFile := false
 	if d.config["pool"] == "" {
 		isFile = !shared.IsDir(srcPath) && !IsBlockdev(srcPath)
+		if strings.HasPrefix(d.config["source"], "cephfs") {
+			//filesystem mount
+			fields := strings.SplitN(d.config["source"], ":", 2)
+			fsName := fields[1]
+			userName := d.config["ceph.user_name"]
+			if userName == "" {
+				userName = "admin"
+			}
+			clusterName := d.config["ceph.cluster_name"]
+			if clusterName == "" {
+				clusterName = "ceph"
+			}
+			path := d.config["path"]
+			err := diskCephfsMount(clusterName, userName, fsName, path)
```

Comment:
So the `cephfs` handling must be done differently. We ultimately need to have the `DiskMount` call do the cephfs mounting.

In order to do so, this part of the logic should only be used to gather the mount options needed for the `ceph` mount call done by `DiskMount`. You'll then combine those mount options with any that may be set in `raw.mount.options`.

`DiskMount` itself needs to be extended to accept a filesystem argument for this to work.

So I'd probably do:
 - Add a new `mntOptions` variable in this function, pre-set it to `d.config["raw.mount.options"]` and change the `DiskMount` call to using that one
 - Add a new `fsName` variable in this function, pre-set it to be an empty string. Add a filesystem name argument to `DIskMount` and have it use the `fsName` variable
 - Rename `diskCephfsMount` to `diskCephfsOptions` only taking in `clusterName`, `userName` and `fsName` and returning a string for the source path of the mount and a string for the mount options.
 - Update `mntOptions` to add the options from `diskCephfsOptions`
 - Set `srcPath` to the mount source path from `diskCephfsOptions`
 - Set `fsName` to `ceph`
 - Update the `srcPath` check to ignore cases where fsName is `ceph`

This should then have `DiskMount` handle all the mounting and the unmount logic should similarly just work.

Reply:
Just to make sure -- the `srcPath` check you're referring to, is this the `IsBlockdev` call in DiskMount on line 68? We want to skip over this check in the instance that `fsName` is ceph? Or are you referring to the error checking that `DiskMount` does after mounting the file system on line 104? (I assumed it wasn't this one, because its part of the core logic of mounting/unmounting)

Reply:
Another question I had -- would a regular `unix.Mount` work for the cephFS, or would it do something similar to mount.ceph like here: (http://manpages.ubuntu.com/manpages/xenial/man8/mount.ceph.8.html) 

I'm guessing it doesn't need to because you had said the regular mount logic works, but I was confused as to why there's an exception here

Reply:
Yeah, I was referring to `	if !shared.PathExists(srcPath) {` which would fail when the `srcPath` is something like `mon:6789:/`.

Using the comma separated syntax for multiple mons sounds good, we should actually switch our `cephfs` storage pool driver to use that as it's more robust in the case where a mon goes offline while the system is in use.